### ISSUE TITLE
Lock congestion in (QNameCache performs terrible under multi-core loads)

### ISSUE THREAD

**Issue by dweiss** [dom4j/dom4j#114]
TL;DR; the parsing performance can be boosted 10x by replacing QNameCache with a non-synchronized, simple version (or even by not caching QNames at all).

The default implementation uses QNameCache extensively to intern and reuse QNames. This cache is implemented using synchronized collections and weak references and performs *terribly* in heavy concurrent scenarios. While I do recognize the interning importance for easing GC activity, the cache should be implemented with the assumption that it's used by a single thread (single reader). Otherwise things go down quickly on multi-core systems.

A longer use-case I just encountered: we have a system that parses terabytes of XMLs, utilizing all available cores (per-thread readers). The performance was about 25MB/s on gzipped input streams on a 64-core machine. The overall system stats showed both the I/O and the CPU idling. When we replaced QNameCache with a non-synchronized version, the performance increased to 150MB/s (compressed input streams so a much higher throughput in reality).

I think the design with synchronized data structures is just not practical these days. The readers should be per-thread and caches should just be thread-local (if any, given how fast GCs are these days).

I attach the two replacement implementations we used for QNameCaches - these are not special at all but do the job:

[QNameNoCache.txt](https://github.com/dom4j/dom4j/files/5526324/QNameNoCache.txt)
[UnsychronizedQNameCache.txt](https://github.com/dom4j/dom4j/files/5526325/UnsychronizedQNameCache.txt)

**Comment by 813221068** [comment]
I have the same problem,Why not use ConcurrentHashMap instead of SynchronizedMap in QNameCache ?

**Comment by FilipJirsak** [comment]
In version 2.2.0, a HashMap in a ThreadLocal is used in QNameCache. In the future, I’m considering the possibility of using different cache implementations, as different approaches are better suited for different types of workloads.
